0012-BIRS-distributed-scheduler.md

```diff
@@ -0,0 +1,611 @@
+  // Otherwise, this is optional.
+  string content_type = 4;
+
+  // HTTP specific fields if request conveys http-compatible request.
+  //
+  // This field is required for http-compatible request. Otherwise,
+  // this field is optional.
+  common.v1.HTTPExtension http_extension = 5;
+}
+
+// JobEventResponse is the response from the app when a job is triggered.
+message JobEventResponse {}
+```
+
+### Scheduler Service
+
+![Pluggable Scheduler Service](./resources/0004-BIRS-distributed-scheduler/pluggableSchedulerService.png)
+
+A new `Scheduler Service` is created in the control plane. This Scheduler Service will include an embedded etcd instance (persisted) as well as a Scheduler Dapr API, which will live in `dapr/proto/scheduler/scheduler.proto`. The Scheduler Service is pluggable and allows for different implementations as needed. It is installed by default into the local development environment on `dapr init` similar to other control plane services in Dapr. This is an optional service and runs in a local container.
+
+To guarantee we don't have several Scheduler Service instances firing off the same job, we will have **virtual partitioning** (in-memory) such that each Scheduler Service instance owns a subset of all the jobs that exist. 
+
+### CLI
+
+`dapr job schedule --name=<name> -–schedule=“@hourly” -–data=“<data>”`
+
+### Implications
+
+- The Scheduler Building Block and Service will result in the ***deprecation*** of `actor reminders` and the `bindings.cron` component.
+
+## Expectations and alternatives
+
+* What is in scope for this proposal?
+    * To start, this will implement generic scheduler logic, then will be expanded to enable the:
+        * Delayed PubSub
+        * Scheduled Service Invocation
+        * Actor Reminders
+* What alternatives have been considered, and why do they not solve the problem?
+    * Placement with Actor reminders is *very* Actor specific. The goal is to have a new scheduler in Dapr that is reusable by several building blocks.
+* What advantages does this proposal have? 
+    * This design has the advantage of enabling the Scheduler Service to be implemented in different ways. 
+    * This design also enables flexibility in where the data is stored. Whether that is in etcd in full, or partially with a reference to a state store (component) of the user's choice.
+
+### Acceptance Criteria
+
+* How will success be measured? 
+    * POCs will be done to guarantee the optimal performance solution for the Scheduler Service, testing with 3 & 5 Scheduler Service instances
+        * minimum RPS for registering reminders
+        * minimum RPS for triggers to app
+        * maximum number of reminders
+        * have a backup and recovery scenario of when Scheduler cluster permanently fails 
+        
+## Completion Checklist
+
+What changes or actions are required to make this proposal complete? Some examples:
+
+* Scheduler Building Block API code
+* Scheduler Service code
+* Tests added (e2e, unit)
```

Comment:
Will this be validated in the weekly/release longhauls as well?

Reply:
Adding new building blocks in alpha does not require longhaul tests. Even to stable has not been mentioned before. I suggest we modify LH tests to use scheduler (with reminders and jobs) as part of the stable criteria, not in alpha.